SQL: Add option to provide the delimiter for the CSV format

bpintea opened the pull request:
This PR adds the option to provide the desired character as the separator
for the CSV format (the default remains comma).
A set of characters are excluded though - like CR, LF, `"` - to avoid
slipping onto the CSV-dialects slope. The tab is also forbidden, the
user needs to choose the `tsv` format explicitely.

Closes #41634.

bpintea committed:
Add option to provide the delimiter to the CSV fmt

This adds the option to provide the desired character as the separator
for the CSV format (the default remains comma).
A set of characters are excluded though - like CR, LF, `"` - to avoid
slipping onto the CSV-dialects slope. The tab is also forbidden, the
user needs to choose the "tsv" format explicitely.

The TSV format also gets a needed supplemental escaping, that of the '\'
character.

bpintea commented:
@elasticmachine run elasticsearch-ci/2

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:
```diff
@@ -109,15 +109,33 @@ protected String eol() {
+        @Override
+        protected void setDelimiter(Character delimiter) {
+            if (delimiter == null) {
+                throw new UnsupportedOperationException();
```
I'd say IllegalArgumentException fits better.

bpintea replied:
Makes sense.

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/TextFormatTests.java:
```diff
@@ -90,7 +92,25 @@ public void testCsvFormatWithRegularData() {
+    }
+
+    public void testCsvFormatWithCustomDelimiterRegularData() {
+        List<Character> forbidden = Arrays.asList('"', '\r', '\n', '\t');
```
Minor, a set would be faster.

bpintea replied:
Changed, thanks.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:
```diff
@@ -124,6 +128,11 @@ public RestResponse buildResponse(SqlQueryResponse response) throws Exception {
 
+    @Override
+    protected Set<String> responseParams() {
+        return textFormat == TextFormat.CSV ? Collections.singleton("delimiter") : Collections.emptySet();
```
Can we use some static variable for "delimiter"?

bpintea replied:
Sure. I've updated along a few other embedded strings, for this change to make sense.

matriv reviewed (commented):
Looks good, left a few comments, but additionally should something be adjusted in docs?

bpintea committed:
Correct delimiter value propagation. Adress comm.

Fix the way the delimiter is being passed around, removing the CSV
member - since its instantiation is a singleton - and pushing it through
function arguments.

The commit fixes a few issues with the "header" URL attribute:
- these was no longer usable, since it wasn't "consumed" by
RestSqlQueryAction before generating the answer (triggering an error
responce on use);
- its value was only checked against "absent", so an incorrect value
(typo like "abaent") could have silently generated an unexpected reply.
- missing

Documentation for the "delimiter" attribute has also been added.

bpintea commented:
> should something be adjusted in docs?

Of course, missed that, thanks for pointing it out.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:
```diff
@@ -235,7 +247,7 @@ String contentType(RestRequest request) {
 
         @Override
-        String maybeEscape(String value) {
+        String maybeEscape(String value, Character __) {
```
Why did you add that second argument? I don't see it being used, or am I missing something :) ?

bpintea replied:
The second param of `maybeEscape`, the delimiter, is only used for the `CSV` value, but not for the `TSV`.

matriv replied:
Of course, sorry, I was just confused.

matriv commented on docs/reference/sql/endpoints/rest.asciidoc:
```diff
@@ -106,6 +106,21 @@ s|Description
+* `delimiter` (default: `,`)
++
+Indicates which character should be used to separate the CSV values. The following values cannot be used for this attribute:
+apostrophe (`"`), carriage-return (`\r`) and new-line (`\n`). The tab (`\t`) can also not be used, the `tsv` format needs to be used
```
Isn't it better to say `double quote`  instead of `apostrophe` ?

bpintea replied:
Oops, you're right, thanks. Fixed.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:
```diff
@@ -186,23 +193,28 @@ String maybeEscape(String value) {
                 return true;
             } else {
-                return !header.toLowerCase(Locale.ROOT).equals("absent");
+                List<String> allowed = List.of(PARAM_HEADER_ABSENT, PARAM_HEADER_PRESENT);
```
You could also use Set here, but personally I would use a `switch` statement on the `head.toLowerCase(Locale.ROOT)` to return true, false, and default => throw the exception.

bpintea replied:
Good suggestion, thanks. Fixed.

matriv reviewed (approved):
LGTM. Left a few more minor comments.

bpintea committed:
Adress review comments. Unescape param

The commit unescapes the deliter param. This is required for the
doc-sourced tests.

Also, implement a switch for param value validation and fix docs.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/csv_delimiter

matriv commented on docs/reference/sql/endpoints/rest.asciidoc:
```diff
@@ -118,7 +118,7 @@ Controls if the CSV header should be included in the response or not. It can tak
 +
 Indicates which character should be used to separate the CSV values. The following values cannot be used for this attribute:
-apostrophe (`"`), carriage-return (`\r`) and new-line (`\n`). The tab (`\t`) can also not be used, the `tsv` format needs to be used
+double quotes (`"`), carriage-return (`\r`) and new-line (`\n`). The tab (`\t`) can also not be used, the `tsv` format needs to be used
```
```suggestion
double quote (`"`), carriage-return (`\r`) and new-line (`\n`). The tab (`\t`) can also not be used, the `tsv` format needs to be used
```
since all the rest are also singular.

matriv reviewed (approved):
LGTM, Thx!
Left one more minor comment in docs.

bpintea committed:
Update docs/reference/sql/endpoints/rest.asciidoc

Co-authored-by: Marios Trivyzas <matriv@gmail.com>

astefan reviewed (commented):
LGTM in general, but I have doubts about adding the `header` request parameter.

bpintea commented on docs/reference/sql/endpoints/rest.asciidoc:
```diff
@@ -106,6 +106,21 @@ s|Description
+Controls if the CSV header should be included in the response or not. It can take one of the following values:
++
+   - `present` (default): include the header;
+   - `absent`: exclude the header.
```
> Why not using a true/false kind of approach with the csv header?

The `absent`/`present` values [were coded in before](https://github.com/elastic/elasticsearch/blob/74c9e5673561eed60700edc5043c5618dad26b2b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java#L186). And my guess is that the values have been chosen to be in-line with the [CT header values](https://github.com/elastic/elasticsearch/blob/74c9e5673561eed60700edc5043c5618dad26b2b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java#L136).

> it was a way of distinguishing between the first page of the response and the subsequent pages of it

I see. ..actually how would a subsequent page be requested with a URL attribute in CSV format? I think we have no `OFFSET`, neither at API, nor at SQL level, do we?
I also found no tests covering this scenario or the use of the `header` attribute.

And these two points above...

> why did you expose this request parameter in the first place?

...let me to believe that it might be an intended feature, but no[t / longer] correctly functioning.

And I think it would be good to have it, CSV handling applications generally offer such functionality. So irrespective of the original intent, is there any concern exposing it?

> Also, the initial request this PR is fixing refers only to the delimiter itself.

True. Would it make sense to split the PR (vs. maybe extending this PR's description)?

astefan replied:
> ...let me to believe that it might be an intended feature, but no[t / longer] correctly functioning.
My 2c on this: if it's requested, we can discuss about adding it. If there is no demand, we don't complicate things.

> I see. ..actually how would a subsequent page be requested with a URL attribute in CSV format? I think we have no OFFSET, neither at API, nor at SQL level, do we?
The first response that comes from ES contains a header called `cursor` to be used for a subsequent call to get the next page of rows. The URL remains the same, but the second call doesn't have a `query` in it, but a `cursor`.

bpintea replied:
> My 2c on this: if it's requested, we can discuss about adding it. If there is no demand, we don't complicate things.

Sure, I can take it out. But just to understand how to take it out:

> it was a way of distinguishing between the first page of the response and the subsequent pages of it

How is this distinguishing being done, since I believe there's no way to specify a `header` URL attribute currently, since it's not "consumed" by the `RestSqlQueryAction` and so using it would generate a `"request [/_sql] contains unrecognized parameter: [header]"` error. Should I remove it entirely then? Sorry if missing something.

> The first response that comes from ES contains a header called cursor to be used for a subsequent call to get the next page of rows.

Right, it's inserted by [`RestSqlQueryAction`](https://github.com/elastic/elasticsearch/blob/380ff0b0d44c0e0036a8934d2b146ee08645ba26/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java#L117), should have spotted it before. Thanks.
Is this documented anywhere though? I couldn't find it on the [REST formats page](https://www.elastic.co/guide/en/elasticsearch/reference/master/sql-rest-format.html). I can add it if not.

astefan replied:
> Sure, I can take it out. But just to understand how to take it out:

Just do not expose it as a request parameter. Leave the rest of the logic unchanged and this PR should only address the delimiter part.

> How is this distinguishing being done, since I believe there's no way to specify a header URL attribute currently, since it's not "consumed" by the RestSqlQueryAction and so using it would generate a "request [/_sql] contains unrecognized parameter: [header]" error. Should I remove it entirely then? Sorry if missing something.

I don't think it's actually being used, indeed. I see logic that's based on the presence of the cursor or not, not much about the `header` being present or not in the request...

> Right, it's inserted by RestSqlQueryAction, should have spotted it before. Thanks.
Is this documented anywhere though? I couldn't find it on the REST formats page. I can add it if not.

It's not documented for other formats (except json and txt), either. I think a simple statement change in [this page](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-pagination.html) also mentioning csv and tsv should be enough (ie, no examples).

bpintea replied:
I've reverted the changes here and opened #60271, since this is clearly a (very minor) bug.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:
```diff
@@ -162,52 +190,58 @@ String maybeEscape(String value) {
         @Override
-        boolean hasHeader(RestRequest request) {
-            String header = request.param("header");
+        boolean needsHeader(RestRequest request) {
```
My interpretation of [what the code does](https://github.com/elastic/elasticsearch/blob/74c9e5673561eed60700edc5043c5618dad26b2b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java#L169) is that it decides if a header is required in the response and not if the RestRequest has a `header` value at all (either URL attribute or HTTP header value). I.e. for the configuration `header=absent` the function will still return `true`.
But I'm happy to revert or apply a more suggestive name if this be an incorrect interpretation.

astefan replied:
The way I looked at it is from the request perspective: the request has a header or not, thus the formatting behavior changes based on this. I would argue further and say that the response needs a header (thus a `needsHeader` method should belong to the SqlQueryResponse object) and not the row-by-row formatting behavior.

bpintea replied:
> thus a `needsHeader` method should belong to the SqlQueryResponse object

Agreed.

I've reverted the change anyways.

bpintea committed:
Address review comments

- rip out anything which isn't Delimiter-related.
- undo 'hasHeader' method renaming.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/csv_delimiter

bpintea committed:
Minor doc update. One more fix revert

Update the doc to make it clear that the textual CSV, TSV and TXT
formats pass the cursor back to the user through the Cursor HTTP header.

bpintea commented:
@elasticmachine update branch

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/csv_delimiter

astefan reviewed (approved):
LGTM
